src/integration-test/java/com/commercetools/sync/integration/services/ProductServiceIT.java

```diff
@@ -136,57 +141,116 @@ public static void tearDown() {
     }
 
     @Test
-    public void cacheKeysToIds_ShouldCacheProductKeysOnlyFirstCall() {
-        Map<String, String> cache = productService.cacheKeysToIds().toCompletableFuture().join();
-        assertThat(cache).hasSize(1);
+    public void getIdFromCacheOrFetch_WithNotCachedExistingProduct_ShouldFetchProductAndCache() {
+        final Optional<String> productId = productService.getIdFromCacheOrFetch(product.getKey())
+                                                         .toCompletableFuture()
+                                                         .join();
+        assertThat(productId).isNotEmpty();
+        assertThat(errorCallBackExceptions).isEmpty();
+        assertThat(errorCallBackMessages).isEmpty();
+    }
 
-        // Create new product without caching
-        final ProductDraft productDraft = createProductDraft(PRODUCT_KEY_2_RESOURCE_PATH, productType.toReference(),
-            null, null, categoryReferencesWithIds,
-            createRandomCategoryOrderHints(categoryReferencesWithIds));
+    @Test
+    public void getIdFromCacheOrFetch_WithNullProductKey_ShouldReturnEmptyOptional() {
+        final Optional<String> productId = productService.getIdFromCacheOrFetch(null)
+                                                         .toCompletableFuture()
+                                                         .join();
+        assertThat(productId).isEmpty();
+        assertThat(errorCallBackExceptions).isEmpty();
+        assertThat(errorCallBackMessages).isEmpty();
+    }
+
+    @Test
+    public void getIdFromCacheOrFetch_WithCachedExistingProduct_ShouldFetchFromCache() {
+        final String oldKey = product.getKey();
+        final Optional<String> oldProductId = productService.getIdFromCacheOrFetch(oldKey)
+                                                            .toCompletableFuture()
+                                                            .join();
+
+        // Change product key on ctp
+        final String newKey = "newKey";
+        productService.updateProduct(product, Collections.singletonList(SetKey.of(newKey)))
+                      .toCompletableFuture()
+                      .join();
+
+        // Fetch product from cache
+        final Optional<String> cachedProductId = productService.getIdFromCacheOrFetch(oldKey)
+                                                               .toCompletableFuture().join();
 
-        CTP_TARGET_CLIENT.execute(ProductCreateCommand.of(productDraft)).toCompletableFuture().join();
+        assertThat(cachedProductId).isNotEmpty();
+        assertThat(cachedProductId).isEqualTo(oldProductId);
+
+        // Fetch product from ctp (because of new key not cached)
+        final Optional<String> productId = productService.getIdFromCacheOrFetch(newKey)
```

Comment:
But actually the test is not testing two things. It's actually testing what is written in it's name: 
`_ShouldFetchFromCache`

What's happening inside is basically:
1. Fetch and cache product id by calling `getIdFromCacheOrFetch`
2. Change product key on CTP.
3. Try to fetch cache again with the same old key using `getIdFromCacheOrFetch`

-> It should fetch the same id because then it means it gets it from the cache, otherwise it won't be able to fetch it from CTP, because this key doesn't exist anymore on CTP.